### PR TITLE
luci-app-tor-hs: add new application

### PR DESCRIPTION
**Pull request by ja-pa**
This PR adds luci app for tor-hs package which helps with creating tor hidden services for router so they can be accessible  throught tor network.

https://github.com/openwrt/packages/tree/master/net/tor-hs

tor-hs-list
tor-hs-config

### CODE REVIEW

File: applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_common.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+'require view';
+'require dom';
+'require uci';
+'require fs';
+'require ui';
+'require rpc';
+'require form';
+'require tools.widgets as widgets';
+
+function setFlagBool(o) {
```
Comment (feckert) [inline applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_common.js:11]: Is this setFlagBool function really needed?
The same applies to the other files

File: applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_common.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	o.enabled = 'true';
+	o.disabled = 'false';
+}
+
+return view.extend({
+	formdata: { torhs: {} },
+
+	callTorHSList: rpc.declare({
```
Comment (feckert) [inline applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_common.js:19]: The common pattern is to define the rpc.declare call not in the view.extend

File: applications/luci-app-tor-hs/root/usr/share/rpcd/acl.d/luci-app-tor-hs.json [hunk 3]
```diff
@@ -0,0 +1,16 @@
+{
+	"luci-app-tor-hs": {
+		"description": "Grant access to Tor hidden service configurator",
+		"read": {
+			"ubus": {
+				"file":["read","list"],
+				"tor_rpcd.sh": ["list-hs"]
```
Comment (feckert) [inline applications/luci-app-tor-hs/root/usr/share/rpcd/acl.d/luci-app-tor-hs.json:7]: I see this rpcd is really named **tor_rpcd.sh** https://github.com/openwrt/packages/blob/master/net/tor-hs/Makefile#L61
but it looks wrong to me. I would suggest that we change that in the package feed. 
@ja-pa Is that really how it's meant to be?

File: applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_status.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+	}),
+
+	load: function() {
+		return Promise.all([
+			this.callTorHSList()
+		]);
+	},
+
+	render: function(data) {
+		var m, s, o,ss;
```
Comment (feckert) [inline applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_status.js:32]: The variable `ss` is not used

File: applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_common.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
+'use strict';
+'require view';
+'require dom';
+'require uci';
+'require fs';
+'require ui';
+'require rpc';
+'require form';
+'require tools.widgets as widgets';
```
Comment (feckert) [inline applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_common.js:9]: Check your require. I don't think they are all required

File: applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_setting.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
+'use strict';
+'require view';
+'require dom';
+'require uci';
+'require fs';
+'require ui';
+'require rpc';
+'require form';
+'require tools.widgets as widgets';
```
Comment (feckert) [inline applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_setting.js:9]: Check your require. I don't think they are all required

File: applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_status.js [hunk 7]
```diff
@@ -0,0 +1,78 @@
+'use strict';
+'require view';
+'require dom';
+'require uci';
+'require fs';
+'require ui';
+'require rpc';
+'require form';
+'require tools.widgets as widgets';
```
Comment (feckert) [inline applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_status.js:9]: Check your require. I don't think they are all required

**Review by feckert — CHANGES_REQUESTED** [review]

File: applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_common.js (hunk 1 above)
Comment (ja-pa) [reply applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_common.js]: No. Removed.

File: applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_common.js (hunk 2 above)
Comment (ja-pa) [reply applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_common.js]: Fixed. Defined outside view.extend

File: applications/luci-app-tor-hs/root/usr/share/rpcd/acl.d/luci-app-tor-hs.json (hunk 3 above)
Comment (ja-pa) [reply applications/luci-app-tor-hs/root/usr/share/rpcd/acl.d/luci-app-tor-hs.json]: Renamed to tor-hs-rpc (https://github.com/openwrt/packages/pull/15097)

File: applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_status.js (hunk 4 above)
Comment (ja-pa) [reply applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_status.js]: Removed

File: applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_status.js (hunk 7 above)
Comment (ja-pa) [reply applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_status.js]: Unnecessary require removed.

File: applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_setting.js (hunk 6 above)
Comment (ja-pa) [reply applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_setting.js]: Unnecessary require removed.

File: applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_common.js (hunk 5 above)
Comment (ja-pa) [reply applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_common.js]: Unnecessary require removed.

**Comment by ja-pa** [comment]
Rebased on master with suggested changes.
@feckert What do you think now ?

**Comment by ja-pa** [comment]
@feckert ping

File: applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_setting.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+		var m, s, o;
+
+		m  = new form.Map('tor-hs', 'Hidden service configuration');
+		m.readonly=false;
+
+		o = m.section(form.TableSection, "hidden-service");
+		o.option(form.Value, "Name", _("Name"));
+		o.option(form.Value, "Description", "Description");
+		o.option(form.Flag, "Enabled", "Enabled");
+		o.option(form.Value, "IPv4", "IPv4");
```
Comment (feckert) [inline applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_setting.js:19]: Wouldn't a datatype check be appropriate here on IPv4?
Translation ?

File: applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_setting.js [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+	render: function(data) {
+		var m, s, o;
+
+		m  = new form.Map('tor-hs', 'Hidden service configuration');
+		m.readonly=false;
+
+		o = m.section(form.TableSection, "hidden-service");
+		o.option(form.Value, "Name", _("Name"));
+		o.option(form.Value, "Description", "Description");
```
Comment (feckert) [inline applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_setting.js:17]: Translation

File: applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_setting.js [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+		o.option(form.Flag, "Enabled", "Enabled");
```
Comment (feckert) [inline applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_setting.js:18]: translation

File: applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_setting.js [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+		o.option(form.Value, "IPv4", "IPv4");
+		o.option(form.DynamicList, "PublicLocalPort", "Public Local Port");
```
Comment (feckert) [inline applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_setting.js:20]: Datatype check Port and translation

File: applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_setting.js [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+		o.option(form.Value, "HookScript", "HookScript");
```
Comment (feckert) [inline applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_setting.js:21]: translation

**Review by feckert — CHANGES_REQUESTED** [review]
I have found a few more things. Which should still be changed.
Thanks

File: applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_common.js [hunk 13]
```diff
@@ -0,0 +1,44 @@
+'use strict';
+'require view';
+'require ui';
+'require rpc';
+'require form';
+
+var callTorHSList = rpc.declare({
+			object: 'tor-hs-rpc',
```
Comment (feckert) [inline applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_common.js:8]: The indentation does not fit. only one tab and the parenthesis at the end of a new line. Please look at other modules how they do this

File: applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_common.js [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+	load: function() {
+		return Promise.all([
+			callTorHSList()
+		]);
+	},
+
+	render: function(data) {
+		var m, s, o;
+
+		m  = new form.Map('tor-hs', 'Common settings');
```
[the remark this answers is not in the document]
Comment (feckert) [reply applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_common.js:24]: ping

File: applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_status.js [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+'require view';
+'require dom';
+'require ui';
+'require rpc';
+'require form';
+
+var callTorHSList = rpc.declare({
+		object: 'tor-hs-rpc',
+		method: 'list-hs',
+		expect: {'':{}  }});
```
Comment (feckert) [inline applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_status.js:11]: The indentation does not fit. only one tab and the parenthesis at the end of a new line. Please look at other modules how they do this

File: applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_status.js [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+
+		s = m.section(form.NamedSection, 'global');
+		s.render = L.bind(function(view, section_id) {
+			return tbl;
+		}, o, this);
+
+		return m.render();
+	},
+
+	addFooter: function() {
```
Comment (feckert) [inline applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_status.js:67]: Why are you overwriting the default function?

File: applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_setting.js [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+		o = m.section(form.TableSection, "hidden-service");
+		o.option(form.Value, "Name", _("Name"));
+		o.option(form.Value, "Description", _("Description"));
+		o.option(form.Flag, "Enabled", _("Enabled"));
+		o.option(form.Value, "IPv4", _("IPv4"));
+		o.datatype = 'ip4addr("nomask")';
+		o.option(form.DynamicList, "PublicLocalPort", _("Public Local Port"));
+		o.option(form.Value, "HookScript", _("HookScript"));
+		o.anonymous = true;
+		o.addremove = true;
```
Comment (feckert) [inline applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_setting.js:24]: Translation should use a `'` and not `"`
String should be also use `'` and not `"` ( for example: `"HookScript"` -> `'HookScript'` )

File: applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_status.js [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+		var uci_hs_list=data[0]["hs-list"];
+		var tbl_lines=[];
+
+		tbl_lines.push(
+				E('tr',{'class':'tr'},[
+				E('th',{'class':'th'} ,_("Name")),
+				E('th',{'class':'th'} ,_("Description")),
+				E('th',{'class':'th'} ,_("Enabled")),
+				E('th',{'class':'th'} ,_("Hidden service address")),
+				E('th',{'class':'th'} ,_("External port"))
```
Comment (feckert) [inline applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_status.js:34]: Translation should use a ' and not "

File: applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_status.js [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+		]));
+
+		if(typeof uci_hs_list  !== 'undefined'){
+			uci_hs_list.forEach(function(entry) {
+				var extern_ports=[]
+				entry.ports.forEach(function(port) {
+					extern_ports.push(port.split(";")[0]);
+				});
```
Comment (feckert) [inline applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_status.js:42]: Please change  `split(";")` to `split(';')`

File: applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_status.js [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+
+				tbl_lines.push(
+					E('tr',{'class':'tr'},[
+						E('td', {'class':'td'}, entry.name),
+						E('td', {'class':'td'}, entry.description),
+						E('td', {'class':'td'}, entry.enabled),
+						E('td', {'class':'td'}, entry.hostname),
+						E('td', {'class':'td'}, extern_ports.join(", "))
```
Comment (feckert) [inline applications/luci-app-tor-hs/htdocs/luci-static/resources/view/tor_hs_status.js:50]: Please change `join(", ")` to `join(', ')`

File: applications/luci-app-tor-hs/root/usr/share/rpcd/acl.d/luci-app-tor-hs.json [hunk 21]
```diff
@@ -0,0 +1,16 @@
...
+				"tor-hs-rpc": ["list-hs"]
+			},
+			"uci": [  "tor-hs" ],
+			"file": {"/etc/tor":["list","read"]}
+		},
+		"write": {
+			"uci": ["tor-hs"]
+		}
+	}
+}
```
Comment (feckert) [inline applications/luci-app-tor-hs/root/usr/share/rpcd/acl.d/luci-app-tor-hs.json:16]: remove one `<space>` before "tor-hs" in read.
Also add `<spaces>` after and before each bracket.

**Review by feckert — CHANGES_REQUESTED** [review]
Sorry this pullrquest slipped through my fingers.
If you change the formal issues then we could merge this.

**Comment by stokito** [comment]
While @ja-pa seems missing and the PR stalled I would like to propose mine PR https://github.com/openwrt/luci/pull/6476#issuecomment-1664804640

**Pull request closed without merging**